### PR TITLE
vm auto-registration: refactoring

### PR DESCRIPTION
**Pull request by yskopets**
**Please provide a description of this PR:**

* refactoring of the VM auto-registration code into separate sub-components
  * logic of connect/disconnect/cleanup lifecycle - `pilot/pkg/workloadentry`
  * logic of health checking - `pilot/pkg/workloadentry/internal/health`
  * logic of persisting internal state as part of `WorkloadEntry` resource - `pilot/pkg/workloadentry/internal/store`
  * logic of auto-registration - `pilot/pkg/workloadentry/internal/autoregistration`
* this PR is merely a refactoring
  * it doesn't add any new features
  * all unit and e2e tests remain unchanged 

* a groundwork for https://github.com/istio/istio/pull/44712

### CODE REVIEW

**Comment by shashankram** [comment]
Are there any functional changes or is it all code reorganization? Could you note this in the PR description?

**Comment by yskopets** [comment]
@shashankram thank you! updated the PR description

File: pilot/pkg/workloadentry/controller.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+var WorkloadControllerAnnotation = workloadentrystore.WorkloadControllerAnnotation
+
+type WorkloadEntryCleaner interface {
+	GetCleanupGracePeriod() time.Duration
+	ShouldCleanup(wle *config.Config, maxConnectionAge time.Duration) bool
+	Cleanup(wle *config.Config, periodic bool)
+}
+
+type Controller struct {
```
Comment (shashankram) [inline pilot/pkg/workloadentry/controller.go:60]: nit: group vars and types together

File: pilot/pkg/workloadentry/controller.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+	maxConnectionAge time.Duration
+
+	wleStore         *workloadentrystore.Controller
+	healthController *health.Controller
+	autoregistration *autoregistration.Controller
+}
+
+// NewController create a controller which manages workload lifecycle and health status.
+func NewController(store model.ConfigStoreController, instanceID string, maxConnAge time.Duration) *Controller {
+	if features.WorkloadEntryAutoRegistration || features.WorkloadEntryHealthChecks {
```
Comment (shashankram) [inline pilot/pkg/workloadentry/controller.go:94]: Consider negating this condition to avoid having the whole function body within the conditional

**Review by shashankram — COMMENTED** [review]
Thanks for clarifying the changes in the PR description

File: pilot/pkg/workloadentry/internal/store/controller.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+// Get retrieves a WorkloadEntry by key.
+func (c *Controller) Get(entryName, entryNs string) *config.Config {
+	return c.store.Get(gvk.WorkloadEntry, entryName, entryNs)
+}
+
+// Create adds a new cWorkloadEntry object to the store.
+func (c *Controller) Create(entry config.Config, conTime time.Time) (revision string, err error) {
+	setConnectMeta(&entry, c.instanceID, conTime)
```
[the remark this answers is not in the document]
Comment (yskopets) [reply pilot/pkg/workloadentry/internal/store/controller.go:171]: Ok, I will rename into `CreateConnected`.

[the remark this answers is not in the document]
Comment (yskopets) [reply pilot/pkg/workloadentry/internal/store/controller.go:171]: > Why does this need to be in it's own "store" package? Things that rely on very specific behavior are beginning to spread out quite a bit.

I've tried to split `autoregistration/controller.go` into more focused sub-components.

I could also leave this code in `autoregistration` but at least move it into a separate file, e.g. `autoregistration/store.go`

File: pilot/pkg/workloadentry/controller.go (hunk 1 above)
Comment (yskopets) [reply pilot/pkg/workloadentry/controller.go:60]: This PR will be replaced by https://github.com/istio/istio/pull/44834.

I've addressed your feedback there.

File: pilot/pkg/workloadentry/controller.go (hunk 2 above)
Comment (yskopets) [reply pilot/pkg/workloadentry/controller.go:94]: This PR will be replaced by https://github.com/istio/istio/pull/44834.

This is the [existing code](https://github.com/istio/istio/blob/619e8c3842b7fcf55bc6addf879e57fecab854e3/pilot/pkg/autoregistration/controller.go#L153). I'm planning to leave it "as is".

File: pilot/pkg/workloadentry/internal/store/controller.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+// Controller knows how to keep internal state as part of a WorkloadEntry resource.
+type Controller struct {
+	instanceID string
+
+	store model.ConfigStoreController
+}
+
+// NewController return a new Controller instance.
+func NewController(store model.ConfigStoreController, instanceID string) *Controller {
```
[the remark this answers is not in the document]
Comment (yskopets) [reply pilot/pkg/workloadentry/internal/store/controller.go:56]: This PR will replaced by https://github.com/istio/istio/pull/44827.

I've renamed the type and function into `Store` in  https://github.com/istio/istio/pull/44827

File: pilot/pkg/workloadentry/internal/store/controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yskopets) [reply pilot/pkg/workloadentry/internal/store/controller.go:171]: This PR will be replaced by https://github.com/istio/istio/pull/44834.

In #44834 I will keep `Create/ChangeStateToConnected/ChangeStateToDisconnected` functions in the `autoregistration/controller.go`.

**Comment by yskopets** [comment]
reworked into https://github.com/istio/istio/pull/44827 and https://github.com/istio/istio/pull/44834

**Pull request closed without merging**